em/emResidualFunctionals.cc

```diff
@@ -0,0 +1,932 @@
+    apf::getIntPoint(me, int_order, i, p);
+    double weight = apf::getIntWeight(me, int_order, i);
+    apf::Matrix3x3 J;
+    apf::getJacobian(me, p, J);
+    double jdet = apf::getJacobianDeterminant(J, dim);
+    w = weight * jdet;
+
+    apf::getVectorShapeValues(el, p, vectorshapes);
+    mth::Matrix<double> vectorShapes (nd, sdim);
+    for (int j = 0; j < nd; j++)
+      for (int k = 0; k < sdim; k++)
+        vectorShapes(j,k) = vectorshapes[j][k];
+
+    mth::Matrix<double> vectorShapesT (sdim, nd);
+    mth::transpose(vectorShapes, vectorShapesT);
+
+    mth::Matrix<double> M (nd,nd);
+    M.zero();
+    mth::multiply(vectorShapes, vectorShapesT, M);
+    M *= w;
+    elmat += M;
+  }
+
+  apf::destroyElement(el);
+  apf::destroyMeshElement(me);
+}
+
+void assembleElementMatrix(apf::Mesh* mesh, apf::MeshEntity*e,
+    apf::Field* f, mth::Matrix<double>& elmat)
+{
+  mth::Matrix<double> curl_elmat, mass_elmat;
+  assembleCurlCurlElementMatrix(mesh, e, f, curl_elmat);
+  assembleVectorMassElementMatrix(mesh, e, f, mass_elmat);
+
+  elmat.resize(curl_elmat.rows(), curl_elmat.cols());
+  elmat.zero();
+  elmat += curl_elmat;
+  elmat += mass_elmat;
+}
+
+/*
+ * computes local bilinear form integral restricted to
+ * an edge of a tet element
+ */
+static double getLocalEdgeBLF(EdgePatch* ep, apf::MeshEntity* tet)
+{
+  PCU_ALWAYS_ASSERT(ep->mesh->getType(tet) == apf::Mesh::TET);
+  // findIn edge in downward edges of tet
+  apf::Downward e;
+  int ne = ep->mesh->getDownward(tet, 1, e);
+  int ei = apf::findIn(e, ne, ep->entity);
+  // get Element Dofs
+  apf::MeshElement* me = apf::createMeshElement(ep->mesh, tet);
+  apf::Element* el = apf::createElement(ep->equilibration->ef, me);
+  int type = ep->mesh->getType(tet);
+  int nd = apf::countElementNodes(el->getFieldShape(), type);
+  apf::NewArray<double> d (nd);
+  el->getElementDofs(d);
+  mth::Vector<double> dofs (nd);
+  for (int i = 0; i < nd; i++)
+    dofs(i) = d[i];
+  // assemble curl curl element matrix
+  mth::Matrix<double> curl_elmat;
+  assembleCurlCurlElementMatrix(ep->mesh, tet,
+      ep->equilibration->ef, curl_elmat);
+  // assemble vector mass element matrix
+  mth::Matrix<double> mass_elmat;
+  assembleVectorMassElementMatrix(ep->mesh, tet,
+      ep->equilibration->ef, mass_elmat);
+  // add element matrices
+  mth::Matrix<double> elmat(nd, nd);
+  elmat.zero();
+  elmat += curl_elmat;
+  elmat += mass_elmat;
+  // multiply element matrix with element dofs
+  mth::Vector<double> blf_integrals (nd);
+  mth::multiply(elmat, dofs, blf_integrals);
+
+  apf::destroyElement(el);
+  apf::destroyMeshElement(me);
+
+  // pick edge index from the resulting vector
+  // negation of negative ND dofs
+  int which, rotate; bool flip;
+  apf::getAlignment(ep->mesh, tet, ep->entity, which, flip, rotate);
+  if (flip)
+    blf_integrals(ei) = -1*blf_integrals(ei);
+  return blf_integrals(ei);
+}
+
+
+// TODO QUESTION redo this to allow user access from outside
```

Comment:
Things that are used for testing should not be here. Remove this function.

Reply:
Resolved. This function was needed here so I have kept it. It is the forcing function.